### PR TITLE
OSGI Interoperability documentation

### PR DESCRIPTION
**Pull request by davidfestal**
Finally !

### CODE REVIEW

File: documentation/1.2/reference/interoperability/osgi.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+# #{page.title_md}
+
+Ceylon is fully interoperable with OSGI, so that Ceylon modules:
+- can be deployed as pure OSGI bundles in an OSGI container out-of-the-box without any modification of the module archive file,
+- can embed additional OSGI metadata, to declare services for example,
+- can easily use OSGI standard services
+
+This provides a great and straightforward opportunity to run Ceylon code inside a growing number of JEE 
+application servers or enterprise containers that are base upon (or integrated with) OSGI.
```
Comment (lucaswerkmeister) [inline documentation/1.2/reference/interoperability/osgi.md]: base**d**

File: documentation/1.2/reference/interoperability/osgi.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+##### Apache Felix 5.4.0:
+
+1. Let's assume we start with a fresh installation of Apache Felix v5.4.0 (`org.apache.felix.main.distribution-5.4.0.zip`)
+
+2. In the `conf/config.properties` file of the Felix installation directory, add the following property:
+
+        org.osgi.framework.executionenvironment=J2SE-1.7,JavaSE-1.7,J2SE-1.6,JavaSE-1.6,J2SE-1.5,JavaSE-1.5,J2SE-1.4,JavaSE-1.4,J2SE-1.3,JavaSE-1.3,J2SE-1.2,,JavaSE-1.2,CDC-1.1/Foundation-1.1,CDC-1.0/Foundation-1.0,J2ME,OSGi/Minimum-1.1,OSGi/Minimum-1.0
+
+  This is necessary since by default the Felix OSGI container provided execution environments don't include `J2SE-1.7`, which is required by a transitive dependency of the `ceylon.net` module.
```
Comment (lucaswerkmeister) [inline documentation/1.2/reference/interoperability/osgi.md]: With the Markdown processor that the website uses, this paragraph is not within the list and therefore breaks the list (`3.` below starts at 1 again). Probably needs to be indented a bit more

File: documentation/1.2/reference/interoperability/osgi.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+#### Installing the Ceylon Distribution and SDK in an OSGI container
+
+##### Apache Felix 5.4.0:
+
+1. Let's assume we start with a fresh installation of Apache Felix v5.4.0 (`org.apache.felix.main.distribution-5.4.0.zip`)
+
+2. In the `conf/config.properties` file of the Felix installation directory, add the following property:
+
+        org.osgi.framework.executionenvironment=J2SE-1.7,JavaSE-1.7,J2SE-1.6,JavaSE-1.6,J2SE-1.5,JavaSE-1.5,J2SE-1.4,JavaSE-1.4,J2SE-1.3,JavaSE-1.3,J2SE-1.2,,JavaSE-1.2,CDC-1.1/Foundation-1.1,CDC-1.0/Foundation-1.0,J2ME,OSGi/Minimum-1.1,OSGi/Minimum-1.0
```
Comment (lucaswerkmeister) [inline documentation/1.2/reference/interoperability/osgi.md]: This line (and other code lines) get “Try Online” links. Not sure what the tag to disable them is…

File: documentation/1.2/reference/interoperability/osgi.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+##### Glassfish v4.1:
+
+Since Glassfish is based on Apache Felix, it can be configured to use the OBR.
+However the simplest way to install the Ceylon Distribution and SDK bundles is to copy then manually.   
+
+Let's assume we start with a fresh installation of Glassfish v4.1
+
+1. Unzip the 2 zip archives mentioned earlier ([distribution](https://downloads.ceylon-lang.org/osgi/distribution/1.2.2/ceylon.distribution.osgi.bundles-1.2.2.zip) and [sdk](https://downloads.ceylon-lang.org/osgi/sdk/1.2.2/ceylon.sdk.osgi.bundles-1.2.2.zip)) into :
+
+  `../glassfish4/glassfish/domains/domain1/autodeploy/bundles`
```
Comment (lucaswerkmeister) [inline documentation/1.2/reference/interoperability/osgi.md]: This also breaks the list (as does the `start-domain` below).

File: documentation/1.2/tour/interop.md [hunk 5]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
 
 ## Deploying Ceylon on OSGi
 
-_TODO: document OSGi interop._
+Ceylon is fully interoperable with OSGI, so that Ceylon modules:
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: This list isn’t a list on the website, probably needs one blank line before the first `-` list item line.

File: documentation/1.2/tour/interop.md [hunk 6]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+- can be deployed as pure OSGI bundles in an OSGI container out-of-the-box without any modification of the module archive file,
+- can embed additional OSGI metadata, to declare services for example,
+- can easily use OSGI standard services
+
+This provides a great and straightforward opportunity to run Ceylon code inside a growing number of JEE application servers or enterprise containers that are base upon (or integrated with) OSGI.
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: base**d** again

File: documentation/1.2/tour/interop.md [hunk 7]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+
+#### Installing the Ceylon distribution / Ceylon SDK in an OSGI container
+
+In order to be able to resolve and start Ceylon module archives (`.car` files) inside an OSGI container, you will first need to install, in the OSGI container, all the bundles of the Ceylon distribution and SDK.
+
+These bundles are available in a dedicated location on the Ceylon language web site under various delivery forms:
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: another broken list

File: documentation/1.2/tour/interop.md [hunk 8]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+The OSGI interoperability reference gives more details about the [URLs providing these packages](../../reference/interoperability/osgi.md#retrieving-the-ceylon-distribution-and-sdk-for-osgi), as well as [how to install](../../reference/interoperability/osgi.md#installing-the-ceylon-distribution-and-sdk-in-an-osgi-container) these bundles, for various containers
+
+#### OSGI Metadata management
+
+[Module archives](./modules#module_archives_and_module_repositories) generated
+for execution on the Java virtual machine already contain the OSGI metadata that
+can be deduced from the Ceylon module descriptor.
+
+More precisely, for the following module descriptor:
+```Ceylon
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: This code block doesn’t work, probably also needs blank lines before and after the ```` ``` ````.

File: documentation/1.2/tour/interop.md [hunk 9]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+Export-Package: example.pureCeylon;version=3.1.0
+Require-Bundle: com.redhat.ceylon.dist;bundle-version=1.2.2;visibility
+ :=reexport,ceylon.locale;bundle-version=1.2.2,ceylon.net;bundle-versi
+ on=1.2.2;visibility:=reexport,ceylon.language;bundle-version=1.2.2;vi
+ sibility:=reexport
+Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version>=1.7))"
+```
+The Ceylon module imports are translated into a `Require-Bundle` header, and `shared` imports are translated into a `reexport` directive.The `java.base "7"` import is translated into a `Require-Capability` header that requires the same Java version.
+
+Note that the `com.redhat.ceylon.dist` bundle is systematically and implicitely added as a dependency to all generated Ceylon archives.
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: “implicitly” without e

File: documentation/1.2/tour/interop.md [hunk 10]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+
+However it is also possible to customize the OSGI metadata of the generated archive in 2 ways:
+
+1. Additional manifest entries or resources (such as a desclarative service descriptor) can be added by simply adding resources as explained [below](./interop.md#meta_inf_and_web_inf).
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: broken link – shouldn’t this be just `#meta_inf_and_web_inf`?

File: documentation/1.2/tour/interop.md [hunk 11]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+  
+  Generally, once omitted from the `Require-Bundle` header, the dependencies to provided bundles should be declared in an `Import-Package` header added in the [`MANIFEST.MF` resource](./interop.md#meta_inf_and_web_inf), as mentioned in point 1.
+ 
+#### Ceylon Metamodel registering
+
+In order to be able to fully leverage the power of the Ceylon language, the metamodel should be initialized for each module used by a Ceylon application. It is automatically performed when running from the command line through the `ceylon run` command, but in an OSGi container it should be done when starting the Ceylon module. The Ceylon OSGI distribution provides an easy way to register the metamodel for a module, as well as for all the modules transitively imported.
+
+This can be done by adding an OSGI activator class to the Ceylon module, that performs this metamodel registration. The `com.redhat.ceylon.dist` OSGI bundle, implicitely required by all Ceylon modules, already provides such an activator class: `com.redhat.ceylon.dist.osgi.Activator`
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: again “implicitly”

File: documentation/1.2/tour/interop.md [hunk 12]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+
+So if you don't need to do anything else in your OSGI Ceylon bundle at startup, you can simply add the following line to your [MANIFEST.MF resource](./interop.md#meta_inf_and_web_inf):
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: this link also seems to be broken

File: documentation/1.2/tour/interop.md [hunk 13]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+```
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: another broken code block

File: documentation/1.2/tour/interop.md [hunk 14]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+Bundle-Activator: com.redhat.ceylon.dist.osgi.Activator
+```
+
+
+Alternatively, if you need to perform some sort initialization in your Ceylon module bundle startup, you can also define an `Activator` class that will delegate to the default `com.redhat.ceylon.dist.osgi.Activator`. In this case:
+
+1. Explicitly import the `com.redhat.ceylon.dist` module in your module descriptor:
+```Ceylon
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: another broken code block (and once it works, it probably needs to be indented so it doesn’t break the list)

File: documentation/1.2/tour/interop.md [hunk 15]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+native("jvm") module example.withActivator "1.0.0" {
+    import com.redhat.ceylon.dist "1.2.2";
+    import java.base "7";
+}
+```
+2. Add the following class to your module:
+```Ceylon
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: broken code block

File: documentation/1.2/tour/interop.md [hunk 16]
```diff
@@ -85,7 +85,107 @@ Aether. You can find more information [here](../../reference/repository/maven).
...
+        super.start(context); // will perform the metamodel registering
+    }
+    shared actual void stop(BundleContext context) {
+        // do module shutdown stuff
+        super.stop(context);
+    }
+}
+```
+3. Add the following line to your `example/withActivator/ROOT/META-INF/MANIFEST.MF` resource:
+```
```
Comment (lucaswerkmeister) [inline documentation/1.2/tour/interop.md]: and another broken code block

**Comment by quintesse** [comment]
There's indeed quite a bit wrong with the layouting. You didn't actually test this, did you? ;)

(I understand, I hate working with this Awestruct crap too)

**Comment by lucaswerkmeister** [comment]
Well, it works on GitHub, and if he can’t build the website, that’s the best substitute…

(I’m working on fixing it, btw)

**Comment by davidfestal** [comment]
@quintesse : I didn't test it in Awestruct yes. Don't have it on my laptop for the moment. But I checked the formatting on the GitHub preview. However I'm also interested in your feelings about the content ;-) do you think it is easy to understand ?

**Comment by quintesse** [comment]
Formatting is quite different unfortunately. If only you could use GitHub to edit those files ;)

I only looked at language , mistakes, etc sorry. Too late for really understanding anything heh

**Comment by lucaswerkmeister** [comment]
Formatting should be fixed if you merge #399 (which updates this PR).

**Comment by davidfestal** [comment]
I merge the fixes provided by Lucas. @gavinking, @FroMage, @quintesse, @tombentley : are you OK with the content, so that it can be published to the live website ?

**Comment by FroMage** [comment]
Done.

**Comment by davidfestal** [comment]
Thanks @FroMage !

**Pull request merged**